libraries/AP_L1_Control/AP_L1_Control.cpp

```diff
@@ -123,8 +123,28 @@ int32_t AP_L1_Control::target_bearing_cd(void) const
  */
 float AP_L1_Control::turn_distance(float wp_radius) const
 {
-    wp_radius *= sq(_ahrs.get_EAS2TAS());
-    return MIN(wp_radius, _L1_dist);
+    float eas2tas_sq = sq(_ahrs.get_EAS2TAS()); // true airspeed correction
+
+    float scaled_turn_distance = wp_radius * eas2tas_sq; // distance scaled to true airspeed
+
+    /** Wind effect i.e. forward crosstrack error correction for the leg to the next waypoint.
+     *  turn early on tail-wind
+     */
+    float headwind = _ahrs.head_wind(); // forward head-wind component in m/s. Negative means tail-wind.
+
+    float arc_distance_for_90_degree_turn_in_meters = 0.5f * 3.14f * wp_radius; // TODO: is there a definition of PI somewhere?
+
+    float ground_speed = _ahrs.groundspeed_vector().length();
+
+    float seconds_needed_for_90_degree_turn = arc_distance_for_90_degree_turn_in_meters / ground_speed;
```

Comment:
I'm not sure assuming constant ground speed is valid. We could probably assume constant turn rate tho...

Reply:
Anything that provides the time needed for the crosswind integral is welcome. Groundspeed was my first guess.

Reply:
IMHO it should be airspeed. If we enter constant rate turn it shouldn't matter if we are being blown 100m downwind, turning 90° should always take T seconds for given radius and airspeed.
Here I made calculation using Desmos for constant rate turn with wind blowing along X axis

R is turn radius
V is airspeed
V_W is wind speed, positive values for tailwind
T is time to turn 360°
T_90 is time to turn 90°
d is total distance traveled downwind
d_w is wind drift component of d
k is T_90 scaling factor to see turns with different angle. Achieved angle is displayed on the row below. 1 for 90° turn.
a_c is lateral acceleration
 
blue line is trajectory over ground 
purple line is exit straight
black line is the exit heading

https://www.desmos.com/calculator/dmugcedoie

Reply:
OK, then we'll go with EAS2TAS(Airspeed) something value.